posts/auditing-using-javers.md

```diff
@@ -0,0 +1,238 @@
+One of the challenges faced in our client is to keep the size of the audited data reasonable. JaVers provide a good domain model configuration, and we were able to use the @ValueObject annotation at the class level to audit on our embedded objects
+
+Example :
+
+```java
+@ValueObject
+public class Address {
+    private final String city;
+    private final String street;
+    private final String zipCode;
+
+    public Address(String city, String street, String zipCode) {
+        this.city = city;
+        this.street = street;
+        this.zipCode = zipCode;
+    }
+
+    // Value Object instances are compared property-by-property,
+    // so the Object.equals() method is ignored by Javers
+    @Override
+    public boolean equals(Object o) {
+        ...
+    }
+}
+```
+
+We also excluded from the audit all the no-PHI data using the @DiffIgnore annotation
+
+Example :
+
+```java
+class MongoStoredEntity {
+
+    // ...
+
+    @DiffIgnore
+    @Property("description")
+    private String description
+
+    // ...
+}
+```
+
+# Retrieve the change
+As mentioned earlier, we stored all our auditing in a separate Mongo collection from client-facing data. However, we needed to find a way to display back the information collected. JaVers provides its' own JaVers Query Language (JQL), which can be used to query the Javers Repository for changes on a given class, object or property.
+
+Data can be retrieved from JaVers in 3 possible ways: Shadows, Changes and Snapshots.
+  * *Shadows* are historical versions of objects
+  * *Change* shows the difference between two objects
+  * *Snapshots* show the state of an object at a point in times
+
+JaVers uses the primary key on the table or collection in order to compile this information.
+
+##Shadows
+Shadows most easily offer a view of data history. Shadows are instances of your object from different points in time. Since Shadows are instances of your domain classes, they can be easily used.
+
+Example:
+```java
+Employee bob = new Employee("bob",
+                             1000,
+                             new Address("London"));
+      javers.commit("author", bob) ;      // initial commit
+
+      bob.setSalary(1200)  ;                // changes
+      bob.getPrimaryAddress().setCity("Paris");  //
+      javers.commit("author", bob);       // second commit
+```
+With the transactions above, we can query the following:
+```java
+Shadows shadows = javers.findShadows(QueryBuilder.byInstance(bob).build());
+```
+Shadows, as provided here is a wrapper object containing all the distinct instances of the Employee object bob.
+
+##Snapshots
+Snapshots are the historical state of a domain object captured as the property-value map. Snapshots are raw data stored in the JaversRepository. When an object is committed, JaVers makes a Snapshot of its state and persists it. JaVers reuses Snapshots and creates a new one only when a given object is changed.
+
+JaVers fetches snapshots in reversed chronological order.
```

Comment:
```suggestion
JaVers fetches snapshots in reverse chronological order.
```

Reply:
Resolved! I would mark it as resolved, but that option isn't showing up anymore.